tests/test_banditPAM.py

```diff
@@ -35,7 +36,32 @@ def test_banditPAM():
     print(kmed2.centers)
 
     # if testing with 'euclidean' distance
-    assert (kmed2.centers == [256, 209, 470, 304])
+    assert np.isin(kmed2.centers, [256, 209, 470, 304]).all()
 
-    # if testing with spearman
-    # assert kmed2.centers == [526, 529, 623, 542]
+
+def test_banditPAM_dask():
+    # load the data
+    ddf = dd.read_csv("tests/banditPAM_data.csv").repartition(npartitions=4)
```

Comment:
For some reason, my machine is unable to read this file while testing. 
I'm getting the following error

```
FAILED tests/test_banditPAM.py::test_banditPAM_dask - TypeError: Cannot interpret '<attribute 'dtype' of 'numpy.generic' objects>' as a data type
```

Reply:
could you give me the full stack trace? I've never gotten that error before

Reply:
sure thing

```
================================================================ test session starts =================================================================
platform darwin -- Python 3.7.10, pytest-6.2.2, py-1.8.2, pluggy-0.13.1
rootdir: /Users/rr/Desktop/code/playground/mark-gam/global-attribution-mapping
plugins: cov-2.11.1
collected 2 items

tests/test_banditPAM.py 	 updated centers -  [4 0 0 0]
	 updated centers -  [  4 584   0   0]
	 updated centers -  [  4 584 245   0]
	 updated centers -  [  4 584 245 328]
	remaining candidates -  2763
	remaining candidates -  1796
	remaining candidates -  772
	remaining candidates -  747
Entering swap with 747 candidates...pts used = 800
Centers after swap -  [657, 584, 245, 328]
	remaining candidates -  2773
	remaining candidates -  2131
	remaining candidates -  1102
	remaining candidates -  691
Entering swap with 691 candidates...pts used = 800
Centers after swap -  [657, 209, 245, 328]
	remaining candidates -  2741
	remaining candidates -  1420
	remaining candidates -  1033
	remaining candidates -  816
Entering swap with 816 candidates...pts used = 800
Centers after swap -  [657, 209, 245, 470]
	remaining candidates -  2627
	remaining candidates -  1559
	remaining candidates -  947
	remaining candidates -  744
Entering swap with 744 candidates...pts used = 800
Centers after swap -  [304, 209, 245, 470]
	remaining candidates -  2784
	remaining candidates -  1972
	remaining candidates -  1010
	remaining candidates -  782
Entering swap with 782 candidates...pts used = 800
Centers after swap -  [304, 209, 256, 470]
	remaining candidates -  2783
	remaining candidates -  1597
	remaining candidates -  937
	remaining candidates -  774
Entering swap with 774 candidates...pts used = 800
	NO Swap -  3 198 0.001615928019154722
Finished test in 9.85
[304, 209, 256, 470]
.F

====================================================================== FAILURES ======================================================================
________________________________________________________________ test_banditPAM_dask _________________________________________________________________

    def test_banditPAM_dask():
        # load the data
>       ddf = dd.read_csv("tests/banditPAM_data.csv").repartition(npartitions=4)

tests/test_banditPAM.py:44:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/lib/python3.7/site-packages/dask/dataframe/io/csv.py:670: in read
    **kwargs,
.venv/lib/python3.7/site-packages/dask/dataframe/io/csv.py:567: in read_pandas
    blocksize=blocksize,
.venv/lib/python3.7/site-packages/dask/dataframe/io/csv.py:306: in text_blocks_to_pandas
    categoricals = head.select_dtypes(include=["category"]).columns
.venv/lib/python3.7/site-packages/pandas/core/frame.py:3427: in select_dtypes
    include_these = Series(not bool(include), index=self.columns)
.venv/lib/python3.7/site-packages/pandas/core/series.py:311: in __init__
    data = sanitize_array(data, index, dtype, copy, raise_cast_failure=True)
.venv/lib/python3.7/site-packages/pandas/core/internals/construction.py:712: in sanitize_array
    subarr = construct_1d_arraylike_from_scalar(value, len(index), dtype)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

value = False, length = 9, dtype = <attribute 'dtype' of 'numpy.generic' objects>

    def construct_1d_arraylike_from_scalar(value, length, dtype):
        """
        create a np.ndarray / pandas type of specified shape and dtype
        filled with values

        Parameters
        ----------
        value : scalar value
        length : int
        dtype : pandas_dtype / np.dtype

        Returns
        -------
        np.ndarray / pandas type of length, filled with value

        """
        if is_datetime64tz_dtype(dtype):
            from pandas import DatetimeIndex

            subarr = DatetimeIndex([value] * length, dtype=dtype)
        elif is_categorical_dtype(dtype):
            from pandas import Categorical

            subarr = Categorical([value] * length, dtype=dtype)
        else:
            if not isinstance(dtype, (np.dtype, type(np.dtype))):
                dtype = dtype.dtype

            if length and is_integer_dtype(dtype) and isna(value):
                # coerce if we have nan for an integer dtype
                dtype = np.dtype("float64")
            elif isinstance(dtype, np.dtype) and dtype.kind in ("U", "S"):
                # we need to coerce to object dtype to avoid
                # to allow numpy to take our string as a scalar value
                dtype = object
                if not isna(value):
                    value = ensure_str(value)

>           subarr = np.empty(length, dtype=dtype)
E           TypeError: Cannot interpret '<attribute 'dtype' of 'numpy.generic' objects>' as a data type

.venv/lib/python3.7/site-packages/pandas/core/dtypes/cast.py:1233: TypeError
================================================================== warnings summary ==================================================================
.venv/lib/python3.7/site-packages/pandas/util/testing.py:27
  /Users/rr/Desktop/code/playground/mark-gam/global-attribution-mapping/.venv/lib/python3.7/site-packages/pandas/util/testing.py:27: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    import pandas._libs.testing as _testing

.venv/lib/python3.7/site-packages/fsspec/__init__.py:47
  /Users/rr/Desktop/code/playground/mark-gam/global-attribution-mapping/.venv/lib/python3.7/site-packages/fsspec/__init__.py:47: DeprecationWarning: SelectableGroups dict interface is deprecated. Use select.
    for spec in entry_points.get("fsspec.specs", []):

.venv/lib/python3.7/site-packages/dask_ml/_typing.py:13
  /Users/rr/Desktop/code/playground/mark-gam/global-attribution-mapping/.venv/lib/python3.7/site-packages/dask_ml/_typing.py:13: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    Number = Union[int, float, np.float, np.int]

.venv/lib/python3.7/site-packages/dask_ml/_typing.py:13
  /Users/rr/Desktop/code/playground/mark-gam/global-attribution-mapping/.venv/lib/python3.7/site-packages/dask_ml/_typing.py:13: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    Number = Union[int, float, np.float, np.int]

.venv/lib/python3.7/site-packages/numba/core/types/__init__.py:108
  /Users/rr/Desktop/code/playground/mark-gam/global-attribution-mapping/.venv/lib/python3.7/site-packages/numba/core/types/__init__.py:108: DeprecationWarning: `np.long` is a deprecated alias for `np.compat.long`. To silence this warning, use `np.compat.long` by itself. In the likely event your code does not need to work on Python 2 you can use the builtin `int` for which `np.compat.long` is itself an alias. Doing this will not modify any behaviour and is safe. When replacing `np.long`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    long_ = _make_signed(np.long)

.venv/lib/python3.7/site-packages/numba/core/types/__init__.py:109
  /Users/rr/Desktop/code/playground/mark-gam/global-attribution-mapping/.venv/lib/python3.7/site-packages/numba/core/types/__init__.py:109: DeprecationWarning: `np.long` is a deprecated alias for `np.compat.long`. To silence this warning, use `np.compat.long` by itself. In the likely event your code does not need to work on Python 2 you can use the builtin `int` for which `np.compat.long` is itself an alias. Doing this will not modify any behaviour and is safe. When replacing `np.long`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    ulong = _make_unsigned(np.long)

tests/test_banditPAM.py::test_banditPAM
tests/test_banditPAM.py::test_banditPAM
tests/test_banditPAM.py::test_banditPAM
tests/test_banditPAM.py::test_banditPAM_dask
tests/test_banditPAM.py::test_banditPAM_dask
tests/test_banditPAM.py::test_banditPAM_dask
  /Users/rr/Desktop/code/playground/mark-gam/global-attribution-mapping/.venv/lib/python3.7/site-packages/pandas/core/indexes/base.py:459: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    elif issubclass(data.dtype.type, np.bool) or is_bool_dtype(data):

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================== short test summary info ===============================================================
FAILED tests/test_banditPAM.py::test_banditPAM_dask - TypeError: Cannot interpret '<attribute 'dtype' of 'numpy.generic' objects>' as a data type
```

Reply:
maybe some sort of dependency mismatch?

Reply:
just jumping in here, but you can probably fix this error by passing in a dict of column names and the expected types.

Reply:
that's what I'm guessing, I'll look into it further

Reply:
ex: `df = dd.read_csv('mylocation.csv', ..., dtype={'ARTICLE_ID': 'object'})`